drv/lpc55-update-server/src/main.rs

```diff
@@ -495,8 +503,45 @@ impl idl::InOrderUpdateImpl for ServerImpl<'_> {
 
         self.image = match (component, slot) {
             (RotComponent::Hubris, SlotId::A)
-            | (RotComponent::Hubris, SlotId::B)
-            | (RotComponent::Stage0, SlotId::B) => Some((component, slot)),
+            | (RotComponent::Hubris, SlotId::B) => {
+                // Fail early on attempt to update the running image.
+                if same_image(component, slot) {
+                    return Err(UpdateError::InvalidSlotIdForOperation.into());
+                }
+
+                // Rollback protection will be implemented by refusing to set
+                // boot preference to an image that has a lower EPOC value in
+                // its caboose.
+                // Setting the boot preference before updating would sidestep that
+                // protection. So, we will fail the prepare step if any
+                // preference settings are selecting the update target image.
+                //
+                // After the update, the boot image selection will be based on:
+                //   - there being only one properly signed image, or
+                //   - transient boot selection (highest priority), or
+                //   - pending persistent selection (altering the persistent
+                //     selection)
+                //   - persistent preference if neither of the above was used.
+
+                let (persistent, pending_persistent, transient) =
+                    self.boot_preferences()?;
```

Comment:
I went back and forth on this but I think since we're now actually checking the transient and persistent settings more places we should make `fn boot_preferences` return a struct e.g.
```
struct BootPreferences{
  persistent: SlotId
  pending_persistent: Option<SlotId>
  transient: Option<SlotId>
}
fn boot_preferences() -> BootPreferences
```
That will make parts of it easier to review